sklearn/cross_decomposition/_pls.py

```diff
@@ -515,6 +446,150 @@ def coef_(self):
     def _more_tags(self):
         return {"poor_score": True, "requires_y": False}
 
+    def _nipals_svd(
+        self,
+        Xk: np.ndarray,
+        Yk: np.ndarray,
+        n_samples: int,
+        n_features: int,
+        n_targets: int,
+        n_components: int,
+    ):
+        """
+        This implements the NIPALS and SVD algorithms from section 4.1 of the
+        Wegelin review paper.
+        """
+        self._norm_y_weights = self.deflation_mode == "canonical"  # 1.1
+        norm_y_weights = self._norm_y_weights
+
+        self.x_weights_ = np.zeros((n_features, n_components))  # U
+        self.y_weights_ = np.zeros((n_targets, n_components))  # V
+        self._x_scores = np.zeros((n_samples, n_components))  # Xi
+        self._y_scores = np.zeros((n_samples, n_components))  # Omega
+        self.x_loadings_ = np.zeros((n_features, n_components))  # Gamma
+        self.y_loadings_ = np.zeros((n_targets, n_components))  # Delta
+        self.n_iter_ = []
+
+        Y_eps = np.finfo(Yk.dtype).eps
+        for k in range(n_components):
+            # Find first left and right singular vectors of the X.T.dot(Y)
+            # cross-covariance matrix.
+            if self.algorithm == "nipals":
+                # Replace columns that are all close to zero with zeros
+                Yk_mask = np.all(np.abs(Yk) < 10 * Y_eps, axis=0)
+                Yk[:, Yk_mask] = 0.0
+
+                try:
+                    (
+                        x_weights,
+                        y_weights,
+                        n_iter_,
+                    ) = _get_first_singular_vectors_power_method(
+                        Xk,
+                        Yk,
+                        mode=self.mode,
+                        max_iter=self.max_iter,
+                        tol=self.tol,
+                        norm_y_weights=norm_y_weights,
+                    )
+                except StopIteration as e:
+                    if str(e) != "Y residual is constant":
+                        raise
```

Comment:
I think we can drop this if / raise branch: `StopIteration` can never be raised when `Y residual` are not constant. This will also make the coverage report happier.